src/Collection/ArrayCollection.php

```diff
@@ -0,0 +1,260 @@
+    {
+        return $this->withModifiers($modifiers);
+    }
+
+    /**
+     * @param ModifierInterface[] $modifiers
+     * @return self
+     */
+    public function withModifiers(array $modifiers): self
+    {
+        $new = clone $this;
+        $new->modifiers = $modifiers;
+        return $new;
+    }
+
+    public function withAddedModifier(ModifierInterface ...$modifiers): self
+    {
+        return $this->withAddedModifiers($modifiers);
+    }
+
+    /**
+     * @param ModifierInterface[] $modifiers
+     * @return self
+     */
+    public function withAddedModifiers(array $modifiers): self
+    {
+        $new = clone $this;
+        $new->modifiers = array_merge($new->modifiers, $modifiers);
+        return $new;
+    }
+
+    /**
+     * @param array|self ...$args
+     * @return self
+     */
+    public function mergeWith(...$args): self
+    {
+        array_unshift($args, $this);
+
+        $arrays = [];
+        foreach ($args as $arg) {
+            $arrays[] = $arg instanceof self ? $arg->data : $arg;
+        }
+
+        $collections = [];
+        foreach ($args as $index => $arg) {
+            $collection = $arg instanceof self ? $arg : new self($arg);
+            foreach ($collection->getModifiers() as $modifier) {
+                if ($modifier instanceof BeforeMergeModifierInterface) {
+                    $collection->data = $modifier->beforeMerge($arrays, $index);
+                }
+            }
+            $collections[$index] = $collection;
+        }
+
+        $collection = $this->merge(...$collections);
+
+        foreach ($collection->getModifiers() as $modifier) {
+            if ($modifier instanceof AfterMergeModifierInterface) {
+                $collection->data = $modifier->afterMerge($collection->data);
+            }
+        }
+
+        return $collection;
+    }
+
+    /**
+     * @param array|self ...$args
+     * @return self
+     */
+    private function merge(...$args): self
+    {
+        $collection = new ArrayCollection();
+
+        while (!empty($args)) {
+            $array = array_shift($args);
+
+            if ($array instanceof ArrayCollection) {
+                $collection->modifiers = array_merge($collection->modifiers, $array->modifiers);
+                $collection->data = $this->merge($collection->data, $array->data)->data;
+                continue;
+            }
+
+            foreach ($array as $k => $v) {
+                if (is_int($k)) {
+                    if (array_key_exists($k, $collection->data)) {
+                        if ($collection->data[$k] !== $v) {
+                            $collection->data[] = $v;
+                        }
+                    } else {
+                        $collection->data[$k] = $v;
+                    }
+                } elseif (
+                    static::isMergable($v) &&
+                    isset($collection->data[$k]) &&
+                    static::isMergable($collection->data[$k])
```

Comment:
```suggestion
                    $this->isMergeable($v) &&
                    isset($collection->data[$k]) &&
                    $this->isMergeable($collection->data[$k])
```

Reply:
isset could be the first in the condition because it's cheaper than function call.